File: sdk/js/src/catalog.ts
import { CoreInterop } from './detail/coreInterop.js';
import { ModelLoadManager } from './detail/modelLoadManager.js';
import { Model } from './detail/model.js';
import { ModelVariant } from './detail/modelVariant.js';
import { ModelInfo } from './types.js';
import { IModel } from './imodel.js';

/**
 * Represents a catalog of AI models available in the system.
 * Provides methods to discover, list, and retrieve models and their variants.
 */
export class Catalog {
    private _name: string;
    private coreInterop: CoreInterop;
    private modelLoadManager: ModelLoadManager;
    private _models: Model[] = [];
    private modelAliasToModel: Map<string, Model> = new Map();
    private modelIdToModelVariant: Map<string, ModelVariant> = new Map();
    private lastFetch: number = 0;
    private updatePromise?: Promise<void>;

    constructor(coreInterop: CoreInterop, modelLoadManager: ModelLoadManager, catalogName?: string) {
        this.coreInterop = coreInterop;
        this.modelLoadManager = modelLoadManager;
        this._name = catalogName ?? this.coreInterop.executeCommand("get_catalog_name");
    }

    /**
     * Gets the name of the catalog.
     * @returns The name of the catalog.
     */
    public get name(): string {
        return this._name;
    }

    /** @internal */
    invalidateCache(): void {
        this.lastFetch = 0;
    }

    private async updateModels(force: boolean = false): Promise<void> {
        // TODO: make this configurable
        // Skip if the cache is still fresh, unless the caller forced a refresh
        // (e.g. self-heal after a cache miss caused by a manually-added BYOM
        // model dropped into the cache directory).
        if (!force && (Date.now() - this.lastFetch) < 6 * 60 * 60 * 1000) { // 6 hours
            return;
        }
        if (this.updatePromise) {
            // If a non-forced refresh is in flight and the caller asked for a
            // forced refresh (e.g. BYOM self-heal), the in-flight result may
            // pre-date the change that prompted the force. Chain a fresh
            // refresh after the current one so the force is not silently
            // dropped — the inner TTL re-check short-circuits if the chained
            // refresh started after our deadline.
            if (force) {
                const chained = this.updatePromise
                    .catch(() => undefined)
                    .then(() => this.runRefreshExclusive());
                return chained;
            }
            return this.updatePromise;
        }
        return this.runRefreshExclusive();
    }

    private runRefreshExclusive(): Promise<void> {
        this.updatePromise = this.fetchAndPopulateModels()
            .finally(() => { this.updatePromise = undefined; });
        return this.updatePromise;
    }

    private async fetchAndPopulateModels(): Promise<void> {
        const modelListJson = await this.coreInterop.executeCommandAsync("get_model_list");
        let modelsInfo: ModelInfo[] = [];
        try {
            modelsInfo = JSON.parse(modelListJson);
        } catch (error) {
            throw new Error(`Failed to parse model list JSON: ${error}`);
        }

        // Incremental refresh: preserve wrapper identity for ids/aliases that
        // survive the refresh so externally-held IModel references keep
        // working with up-to-date metadata and (for Model) keep any explicit
        // selectVariant() choice. New ids get fresh wrappers; removed ids get
        // evicted.

        const freshIds = new Set<string>();
        const freshAliasGroups = new Map<string, ModelInfo[]>();
        for (const info of modelsInfo) {
            freshIds.add(info.id);
            const group = freshAliasGroups.get(info.alias);
            if (group) {
                group.push(info);
            } else {
                freshAliasGroups.set(info.alias, [info]);
            }
        }

        for (const staleId of [...this.modelIdToModelVariant.keys()]) {
            if (!freshIds.has(staleId)) {
                this.modelIdToModelVariant.delete(staleId);
            }
        }
        for (const staleAlias of [...this.modelAliasToModel.keys()]) {
            if (!freshAliasGroups.has(staleAlias)) {
                this.modelAliasToModel.delete(staleAlias);
            }
        }

        for (const info of modelsInfo) {
            const existing = this.modelIdToModelVariant.get(info.id);
            if (existing) {
                existing._refreshInfo(info);
            } else {
                this.modelIdToModelVariant.set(
                    info.id,
                    new ModelVariant(info, this.coreInterop, this.modelLoadManager)
                );
            }
        }

        const refreshedAliasOrder: Model[] = [];
        for (const [alias, infos] of freshAliasGroups) {
            const variants = infos.map(i => this.modelIdToModelVariant.get(i.id)!);
            const existingModel = this.modelAliasToModel.get(alias);
            if (existingModel) {
                existingModel._refreshVariants(variants);
                refreshedAliasOrder.push(existingModel);
            } else {
                const m = new Model(variants[0]);
                for (let i = 1; i < variants.length; i++) {
                    m.addVariant(variants[i]);
                }
                this.modelAliasToModel.set(alias, m);
                refreshedAliasOrder.push(m);
            }
        }

        this._models = refreshedAliasOrder;
        this.lastFetch = Date.now();
    }

    /**
     * Lists all available models in the catalog.
     * This method is asynchronous as it may fetch the model list from a remote service or perform file I/O.
     * @returns A Promise that resolves to an array of IModel objects.
     */
    public async getModels(): Promise<IModel[]> {
        await this.updateModels();
        return this._models;
    }

    /**
     * Retrieves a model by its alias.
     * This method is asynchronous as it may ensure the catalog is up-to-date by fetching from a remote service.
     * @param alias - The alias of the model to retrieve.
     * @returns A Promise that resolves to the IModel object if found, otherwise throws an error.
     * @throws Error - If alias is null, undefined, or empty.
     */
    public async getModel(alias: string): Promise<IModel> {
        if (typeof alias !== 'string' || alias.trim() === '') {
            throw new Error('Model alias must be a non-empty string.');
        }
        await this.updateModels();
        let model = this.modelAliasToModel.get(alias);
        if (!model) {
            // Self-heal: the alias may belong to a BYOM model added to the
            // cache directory after our last catalog refresh.
            await this.updateModels(true);
            model = this.modelAliasToModel.get(alias);
        }
        if (!model) {
            const availableAliases = Array.from(this.modelAliasToModel.keys()).join(', ');
            throw new Error(`Model with alias '${alias}' not found. Available models: ${availableAliases || '(none)'}`);
        }
        return model;
    }

    /**
     * Retrieves a specific model variant by its ID.
     * NOTE: This will return an IModel with a single variant. Use getModel to get an IModel with all available
     * variants.
     * This method is asynchronous as it may ensure the catalog is up-to-date by fetching from a remote service.
     * @param modelId - The unique identifier of the model variant.
     * @returns A Promise that resolves to the IModel object if found, otherwise throws an error.
     * @throws Error - If modelId is null, undefined, or empty.
     */
    public async getModelVariant(modelId: string): Promise<IModel> {
        if (typeof modelId !== 'string' || modelId.trim() === '') {
            throw new Error('Model ID must be a non-empty string.');
        }
        await this.updateModels();
        let variant = this.modelIdToModelVariant.get(modelId);
        if (!variant) {
            // Self-heal: the id may belong to a BYOM model added to the cache
            // directory after our last catalog refresh.
            await this.updateModels(true);
            variant = this.modelIdToModelVariant.get(modelId);
        }
        if (!variant) {
            const availableIds = Array.from(this.modelIdToModelVariant.keys()).join(', ');
            throw new Error(`Model variant with ID '${modelId}' not found. Available variants: ${availableIds || '(none)'}`);
        }
        return variant;
    }

    /**
     * Retrieves a list of all locally cached model variants.
     * This method is asynchronous as it may involve file I/O or querying the underlying core.
     * @returns A Promise that resolves to an array of cached IModel objects.
     */
    public async getCachedModels(): Promise<IModel[]> {
        await this.updateModels();
        const cachedModelListJson = await this.coreInterop.executeCommandAsync("get_cached_models");
        let cachedModelIds: string[] = [];
        try {
            cachedModelIds = JSON.parse(cachedModelListJson);
        } catch (error) {
            throw new Error(`Failed to parse cached model list JSON: ${error}`);
        }
        return this.resolveModelIds(cachedModelIds);
    }

    /**
     * Retrieves a list of all currently loaded model variants.
     * This operation is asynchronous because checking the loaded status may involve querying
     * the underlying core or an external service, which can be an I/O bound operation.
     * @returns A Promise that resolves to an array of loaded IModel objects.
     */
    public async getLoadedModels(): Promise<IModel[]> {
        await this.updateModels();
        let loadedModelIds: string[] = [];
        try {
            loadedModelIds = await this.modelLoadManager.listLoaded();
        } catch (error) {
            throw new Error(`Failed to list loaded models: ${error}`);
        }
        return this.resolveModelIds(loadedModelIds);
    }

    /**
     * Resolve a list of model ids against the in-memory catalog, self-healing once
     * if any id is unknown (e.g. a manually-added BYOM model the SDK has not yet seen).
     * Preserves the input order of `modelIds` (minus unknowns), deduplicating variants.
     */
    private async resolveModelIds(modelIds: string[]): Promise<IModel[]> {
        if (modelIds.some(id => !this.modelIdToModelVariant.has(id))) {
            await this.updateModels(true);
        }

        const resolved: IModel[] = [];
        const seen = new Set<IModel>();
        for (const modelId of modelIds) {
            const variant = this.modelIdToModelVariant.get(modelId);
            if (variant && !seen.has(variant)) {
                resolved.push(variant);
                seen.add(variant);
            }
        }
        return resolved;
    }

    /**
     * Get the latest version of a model.
     * This is used to check if a newer version of a model is available in the catalog for download.
     * @param modelOrModelVariant - The model to check for the latest version.
     * @returns The latest version of the model. Will match the input if it is the latest version.
     */
    public async getLatestVersion(modelOrModelVariant: IModel): Promise<IModel> {
        await this.updateModels();

        // Resolve to the parent Model by alias
        const model = this.modelAliasToModel.get(modelOrModelVariant.alias);
        if (!model) {
            throw new Error(`Model with alias '${modelOrModelVariant.alias}' not found in catalog.`);
        }

        // variants are sorted by version, so the first one matching the name is the latest version
        const latest = model.variants.find(v => v.info.name === modelOrModelVariant.info.name);
        if (!latest) {
            throw new Error(
                `Internal error. Mismatch between model (alias:${model.alias}) and ` +
                `model variant (alias:${modelOrModelVariant.alias}).`
            );
        }

        // if input was the latest return the input (could be model or model variant)
        // otherwise return the latest model variant
        return latest.id === modelOrModelVariant.id ? modelOrModelVariant : latest;
    }
}

File: sdk/cs/src/Catalog.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Microsoft.AI.Foundry.Local;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

using Microsoft.AI.Foundry.Local.Detail;
using Microsoft.Extensions.Logging;

internal sealed class Catalog : ICatalog, IDisposable
{
    private readonly Dictionary<string, Model> _modelAliasToModel = new();
    private readonly Dictionary<string, ModelVariant> _modelIdToModelVariant = new();
    private DateTime _lastFetch;

    private readonly IModelLoadManager _modelLoadManager;
    private readonly ICoreInterop _coreInterop;
    private readonly ILogger _logger;
    private readonly AsyncLock _lock = new();

    public string Name { get; init; }

    private Catalog(IModelLoadManager modelLoadManager, ICoreInterop coreInterop, ILogger logger)
    {
        _modelLoadManager = modelLoadManager;
        _coreInterop = coreInterop;
        _logger = logger;

        _lastFetch = DateTime.MinValue;

        CoreInteropRequest? input = null;
        var response = coreInterop.ExecuteCommand("get_catalog_name", input);
        if (response.Error != null)
        {
            throw new FoundryLocalException($"Error getting catalog name: {response.Error}", _logger);
        }

        Name = response.Data!;
    }

    internal static async Task<Catalog> CreateAsync(IModelLoadManager modelManager, ICoreInterop coreInterop,
                                                    ILogger logger, CancellationToken? ct = null)
    {
        var catalog = new Catalog(modelManager, coreInterop, logger);
        await catalog.UpdateModels(ct).ConfigureAwait(false);
        return catalog;
    }

    public async Task<List<IModel>> ListModelsAsync(CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => ListModelsImplAsync(ct),
                                                     "Error listing models.", _logger).ConfigureAwait(false);
    }

    public async Task<List<IModel>> GetCachedModelsAsync(CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => GetCachedModelsImplAsync(ct),
                                                     "Error getting cached models.", _logger).ConfigureAwait(false);
    }

    public async Task<List<IModel>> GetLoadedModelsAsync(CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => GetLoadedModelsImplAsync(ct),
                                                     "Error getting loaded models.", _logger).ConfigureAwait(false);
    }

    public async Task<IModel?> GetModelAsync(string modelAlias, CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => GetModelImplAsync(modelAlias, ct),
                                                     $"Error getting model with alias '{modelAlias}'.", _logger)
                                                    .ConfigureAwait(false);
    }

    public async Task<IModel?> GetModelVariantAsync(string modelId, CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => GetModelVariantImplAsync(modelId, ct),
                                                     $"Error getting model variant with ID '{modelId}'.", _logger)
                                                    .ConfigureAwait(false);
    }

    public async Task<IModel> GetLatestVersionAsync(IModel modelOrModelVariant, CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(
            () => GetLatestVersionImplAsync(modelOrModelVariant, ct),
            $"Error getting latest version for model with name '{modelOrModelVariant.Info.Name}'.",
            _logger).ConfigureAwait(false);
    }

    private async Task<List<IModel>> ListModelsImplAsync(CancellationToken? ct = null)
    {
        await UpdateModels(ct).ConfigureAwait(false);

        using var disposable = await _lock.LockAsync().ConfigureAwait(false);
        return _modelAliasToModel.Values.OrderBy(m => m.Alias).Cast<IModel>().ToList();
    }

    private async Task<List<IModel>> GetCachedModelsImplAsync(CancellationToken? ct = null)
    {
        await UpdateModels(ct).ConfigureAwait(false);

        var cachedModelIds = await Utils.GetCachedModelIdsAsync(_coreInterop, ct).ConfigureAwait(false);
        return await ResolveModelIdsAsync(cachedModelIds, ct).ConfigureAwait(false);
    }

    private async Task<List<IModel>> GetLoadedModelsImplAsync(CancellationToken? ct = null)
    {
        await UpdateModels(ct).ConfigureAwait(false);

        var loadedModelIds = await _modelLoadManager.ListLoadedModelsAsync(ct).ConfigureAwait(false);
        return await ResolveModelIdsAsync(loadedModelIds, ct).ConfigureAwait(false);
    }

    // Resolve a list of model ids against the in-memory catalog, self-healing once
    // if any id is unknown (e.g. a manually-added BYOM model the SDK has not yet seen).
    // Preserves the input order of modelIds (minus unknowns).
    private async Task<List<IModel>> ResolveModelIdsAsync(string[] modelIds, CancellationToken? ct)
    {
        bool needsRefresh = false;
        using (var disposable = await _lock.LockAsync().ConfigureAwait(false))
        {
            foreach (var modelId in modelIds)
            {
                if (!_modelIdToModelVariant.ContainsKey(modelId))
                {
                    needsRefresh = true;
                    break;
                }
            }
        }

        if (needsRefresh)
        {
            await UpdateModels(ct, force: true).ConfigureAwait(false);
        }

        List<IModel> resolved = new(modelIds.Length);
        using (var disposable = await _lock.LockAsync().ConfigureAwait(false))
        {
            foreach (var modelId in modelIds)
            {
                if (_modelIdToModelVariant.TryGetValue(modelId, out ModelVariant? variant))
                {
                    resolved.Add(variant);
                }
            }
        }

        return resolved;
    }

    private async Task<Model?> GetModelImplAsync(string modelAlias, CancellationToken? ct = null)
    {
        if (string.IsNullOrWhiteSpace(modelAlias))
        {
            return null;
        }

        await UpdateModels(ct).ConfigureAwait(false);

        Model? model;
        using (var disposable = await _lock.LockAsync().ConfigureAwait(false))
        {
            _modelAliasToModel.TryGetValue(modelAlias, out model);
        }
        if (model is not null)
        {
            return model;
        }

        // Self-heal: the alias may belong to a BYOM model added to the cache
        // directory after our last catalog refresh.
        await UpdateModels(ct, force: true).ConfigureAwait(false);
        using (var disposable = await _lock.LockAsync().ConfigureAwait(false))
        {
            _modelAliasToModel.TryGetValue(modelAlias, out model);
        }
        return model;
    }

    private async Task<ModelVariant?> GetModelVariantImplAsync(string modelId, CancellationToken? ct = null)
    {
        if (string.IsNullOrWhiteSpace(modelId))
        {
            return null;
        }

        await UpdateModels(ct).ConfigureAwait(false);

        ModelVariant? modelVariant;
        using (var disposable = await _lock.LockAsync().ConfigureAwait(false))
        {
            _modelIdToModelVariant.TryGetValue(modelId, out modelVariant);
        }
        if (modelVariant is not null)
        {
            return modelVariant;
        }

        // Self-heal: the id may belong to a BYOM model added to the cache
        // directory after our last catalog refresh.
        await UpdateModels(ct, force: true).ConfigureAwait(false);
        using (var disposable = await _lock.LockAsync().ConfigureAwait(false))
        {
            _modelIdToModelVariant.TryGetValue(modelId, out modelVariant);
        }
        return modelVariant;
    }

    private async Task<IModel> GetLatestVersionImplAsync(IModel modelOrModelVariant, CancellationToken? ct)
    {
        Model? model;

        if (modelOrModelVariant is ModelVariant)
        {
            // For ModelVariant, resolve the owning Model via alias.
            model = await GetModelImplAsync(modelOrModelVariant.Alias, ct);
        }
        else
        {
            // Try to use the concrete Model instance if this is our SDK type.
            model = modelOrModelVariant as Model;

            // If this is a different IModel implementation (e.g., a test stub),
            // fall back to resolving the Model via alias.
            if (model == null)
            {
                model = await GetModelImplAsync(modelOrModelVariant.Alias, ct);
            }
        }

        if (model == null)
        {
            throw new FoundryLocalException($"Model with alias '{modelOrModelVariant.Alias}' not found in catalog.",
                                            _logger);
        }

        // variants are sorted by version, so the first one matching the name is the latest version for that variant.
        var latest = model!.Variants.FirstOrDefault(v => v.Info.Name == modelOrModelVariant.Info.Name) ??
            // should not be possible given we internally manage all the state involved
            throw new FoundryLocalException($"Internal error. Mismatch between model (alias:{model.Alias}) and " +
                                            $"model variant (alias:{modelOrModelVariant.Alias}).", _logger);

        // if input was the latest return the input (could be model or model variant)
        // otherwise return the latest model variant
        return latest.Id == modelOrModelVariant.Id ? modelOrModelVariant : latest;
    }

    private async Task UpdateModels(CancellationToken? ct, bool force = false)
    {
        // TODO: make this configurable
        // Skip if the cache is still fresh, unless the caller forced a refresh
        // (e.g. self-heal after a cache miss caused by a manually-added BYOM
        // model dropped into the cache directory).
        if (!force && DateTime.Now - _lastFetch < TimeSpan.FromHours(6))
        {
            return;
        }

        CoreInteropRequest? input = null;
        var result = await _coreInterop.ExecuteCommandAsync("get_model_list", input, ct).ConfigureAwait(false);

        if (result.Error != null)
        {
            throw new FoundryLocalException($"Error getting models: {result.Error}", _logger);
        }

        var models = JsonSerializer.Deserialize(result.Data!, JsonSerializationContext.Default.ListModelInfo);
        if (models == null)
        {
            _logger.LogDebug($"ListModelInfo deserialization error in UpdateModels. Data: {result.Data}");
            throw new FoundryLocalException($"Failed to deserialize models from response.", _logger);
        }

        using var disposable = await _lock.LockAsync().ConfigureAwait(false);

        // Incremental refresh: preserve wrapper identity for ids/aliases that
        // survive the refresh so externally-held IModel references keep
        // working with up-to-date metadata and (for Model) keep any explicit
        // SelectVariant() choice. New ids get fresh wrappers; removed ids get
        // evicted.

        var freshIds = new HashSet<string>(StringComparer.Ordinal);
        var freshAliasGroups = new Dictionary<string, List<ModelInfo>>(StringComparer.Ordinal);
        foreach (var info in models)
        {
            freshIds.Add(info.Id);
            if (!freshAliasGroups.TryGetValue(info.Alias, out var group))
            {
                group = [];
                freshAliasGroups[info.Alias] = group;
            }
            group.Add(info);
        }

        foreach (var staleId in _modelIdToModelVariant.Keys.Where(id => !freshIds.Contains(id)).ToList())
        {
            _modelIdToModelVariant.Remove(staleId);
        }
        foreach (var staleAlias in _modelAliasToModel.Keys.Where(a => !freshAliasGroups.ContainsKey(a)).ToList())
        {
            _modelAliasToModel.Remove(staleAlias);
        }

        foreach (var info in models)
        {
            if (_modelIdToModelVariant.TryGetValue(info.Id, out var existing))
            {
                existing.RefreshInfo(info);
            }
            else
            {
                _modelIdToModelVariant[info.Id] = new ModelVariant(info, _modelLoadManager, _coreInterop, _logger);
            }
        }

        foreach (var kvp in freshAliasGroups)
        {
            var alias = kvp.Key;
            var aliasInfos = kvp.Value;
            var aliasVariants = aliasInfos.ConvertAll(i => _modelIdToModelVariant[i.Id]);
            if (_modelAliasToModel.TryGetValue(alias, out var existingModel))
            {
                existingModel.RefreshVariants(aliasVariants);
            }
            else
            {
                var m = new Model(aliasVariants[0], _logger);
                for (var i = 1; i < aliasVariants.Count; i++)
                {
                    m.AddVariant(aliasVariants[i]);
                }
                _modelAliasToModel[alias] = m;
            }
        }

        _lastFetch = DateTime.Now;
    }

    internal void InvalidateCache()
    {
        _lastFetch = DateTime.MinValue;
    }

    public void Dispose()
    {
        _lock.Dispose();
    }
}
